reconcile/prometheus_rules_tester/integration.py

```diff
@@ -0,0 +1,266 @@
+    """Check that all services in Prometheus rules are known.
+    This replaces an enum in the json schema with a list
+    in app-interface settings."""
+    allowed_services = settings["alertingServices"]
+    missing_services = set()
+    spec = rule["spec"]
+    groups = spec["groups"]
+    for g in groups:
+        group_rules = g["rules"]
+        for r in group_rules:
+            rule_labels = r.get("labels")
+            if not rule_labels:
+                continue
+            service = rule_labels.get("service")
+            if not service:
+                continue
+            if service not in allowed_services:
+                missing_services.add(service)
+
+    if missing_services:
+        return CommandExecutionResult(
+            False, f"services are missing from alertingServices: {missing_services}"
+        )
+
+    return CommandExecutionResult(True, "")
+
+
+# We return here a CommandExecutionResult as it is what run_test function has to
+# add to the "result" field.
+def check_rule_length(rule_length: int) -> CommandExecutionResult:
+    """Checks rule length so that prom operator has no issues with it"""
+    if rule_length > MAX_CONFIGMAP_SIZE:
+        return CommandExecutionResult(
+            False, f"Rules spec is larger than {MAX_CONFIGMAP_SIZE} bytes."
+        )
+    return CommandExecutionResult(True, "")
+
+
+def run_test(test: Test, settings: Mapping) -> None:
+    """Checks rules, run tests and stores the result in test.result"""
+    check_rule_result = promtool.check_rule(test.rule["spec"])
+    valid_services_result = check_valid_services(test.rule, settings)
+    rule_length_result = check_rule_length(test.rule_length)
+    test.result = check_rule_result and valid_services_result and rule_length_result
+
+    if not test.result:
+        return
+
+    rule_files = {test.rule_path: test.rule["spec"]}
+    for t in test.tests or []:
+        result = promtool.run_test(t.test, rule_files)
+        test.result = test.result and result
+
+
+def check_rules_and_tests(
+    settings: Mapping, thread_pool_size: int, cluster_name: Optional[str] = None
+) -> list[Test]:
+    """Fetch rules and associated tests, run checks on rules and tests if they exist
+    and return a list of failed checks/tests"""
+    tests = get_rules_and_tests(cluster_name, settings, thread_pool_size)
+    threaded.run(
+        func=run_test,
+        iterable=tests,
+        thread_pool_size=thread_pool_size,
+        settings=settings,
+    )
+
+    failed_tests = [test for test in tests if not test.result]
+
+    return failed_tests
+
+
+def run(
+    dry_run: bool, thread_pool_size: int = 5, cluster_name: Optional[str] = None
+) -> None:
+    """Check prometheus rules syntax and run the tests associated to them"""
+    orb.QONTRACT_INTEGRATION = QONTRACT_INTEGRATION
+    orb.QONTRACT_INTEGRATION_VERSION = QONTRACT_INTEGRATION_VERSION
+
+    settings = queries.get_app_interface_settings()
```

Comment:
Same, can we get these from a typed query?

Reply:
Indeed, I will change it.

Reply:
fixed in 2d04214b